### PR TITLE
[FLINK-33121] Failed precondition in JobExceptionsHandler due to concurrent global failures

### PR DESCRIPTION
**Pull request by pgaref**
## What is the purpose of the change

We make the assumption that Global Failures (with null Task name) may only be RootExceptions and and Local/Task exception may be part of concurrent exceptions List (see JobExceptionsHandler#createRootExceptionInfo). 
However, when the Adaptive scheduler is in a Restarting phase due to an existing failure (that is now the new Root) we can still, in rare occasions, capture new Global failures, violating this condition (with an assertion is thrown as part of assertLocalExceptionInfo).

In this PR we ignore Global failures while being in a Restarting phase on the Adaptive scheduler until we properly support multiple Global failures in the Exception History as part of https://issues.apache.org/jira/browse/FLINK-34922

Note: DefaultScheduler does not suffer from this issue as it treats failures directly as HistoryEntries (no conversion step)

https://issues.apache.org/jira/browse/FLINK-33121

## Brief change log

  - Ignore Global failures while being in a Restarting phase on the Adaptive scheduler 

## Verifying this change

This change added tests and can be verified as follows:

  - Extended AdaptiveSchedulerITCase with `testGlobalFailureOnRestart`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

**Comment by pgaref** [comment]
> I'm against making this change without a clear explanation as to when this case occurs. AFAICT we don't intend for it to occur, so let's find that bug rather than allowing it and potentially causing other downstream side-effects.

Found the root cause of this, @zentol / @dmvk  can you please take another look?

**Review by dmvk — CHANGES_REQUESTED** [review]
The reasoning reads sane 👍 I have few questions
1. Would it be possible to introduce a test that simulates the original issue?
2. What about `testExceptionHistoryWithTaskConcurrentGlobalFailure`, which is failing? 
  a. It suggests multiple global failures can occur. Is that even correct?
  b. If correct, wouldn't that manifest into the same problem? (we can only have one root I assume)
3. Is the wrong ordering a regression caused by failure enrichment?

**Comment by pgaref** [comment]
Hey @dmvk  -- thanks for the comments!

1. Was debating about that, however testing e2e would require to add the mocked AdaptiveScheduler dependency (like `ExceptionHistoryTester`) to JobExceptionsHandlerTest -- that even though is possible, would add more complexity than value IMO as we just need to ensure the order of entries for the conversion is correct (similar to what we do in the AdaptiveSchedulerTest in the PR)
2. Regarding `testExceptionHistoryWithTaskConcurrentGlobalFailure`, two concurrent Global failures could never occur with the current logic in the AdaptiveScheduler -- so I modified the test to represent reality and include our case too
3. This issue is not related to failure enrichment at all -- it is just not too common as there should be one or more concurrent Task/Local failures followed by a Global one

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/AdaptiveSchedulerTest.java [hunk 1]
```diff
@@ -1619,12 +1619,18 @@ void testExceptionHistoryWithTaskFailureFromStopWithSavepoint() throws Exception
 
     @Test
     void testExceptionHistoryWithTaskConcurrentGlobalFailure() throws Exception {
```
Comment (dmvk) [inline flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/AdaptiveSchedulerTest.java:1621]: is the test name still matching what's happening in the test?

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/StateWithExecutionGraph.java [hunk 2]
```diff
@@ -389,7 +389,10 @@ CompletableFuture<CoordinationResponse> deliverCoordinationRequestToCoordinator(
     @Override
     public void handleGlobalFailure(
             Throwable cause, CompletableFuture<Map<String, String>> failureLabels) {
-        failureCollection.add(ExceptionHistoryEntry.createGlobal(cause, failureLabels));
+        // Ignore Global failures when restarting to avoid messing up the failure history
+        if (this.getJob().getState() != JobStatus.RESTARTING) {
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/StateWithExecutionGraph.java:393]: Override `handleGlobalFailure` in `Restarting` instead.

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/exceptionhistory/RootExceptionHistoryEntry.java [hunk 3]
```diff
@@ -104,7 +104,7 @@ public static RootExceptionHistoryEntry fromExceptionHistoryEntry(
                 entry.getException(),
                 entry.getTimestamp(),
                 entry.getFailureLabelsFuture(),
-                null,
+                entry.getFailingTaskName(),
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/exceptionhistory/RootExceptionHistoryEntry.java:107]: seems unrelated?

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/StateWithExecutionGraph.java (hunk 2 above)
Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/StateWithExecutionGraph.java:393]: While less common this problem can also occur for `Canceling` and `Failing`.

**Review by zentol — CHANGES_REQUESTED** [review]
I suppose this can work but it seems rather brittle and may obfuscate the underlying failure cause.
If the reality is that multiple global failures can occur then the exception history should simply support that. The old stack got around this by just throwing away older global failures if a new one occurred, which isn't great either.

This might be the opportunity to properly fix this. In the old days (pre source coordinators) this wasn't as pressing because global failures were rare and only triggered by the JM (which typically ensured it only happened once), but since this has changed now we should adjust accordingly.

**Comment by pgaref** [comment]
> I suppose this can work but it seems rather brittle and may obfuscate the underlying failure cause. If the reality is that multiple global failures can occur or that global failures can occur after task failures then the exception history should simply support that. The old stack got around this by just throwing away older global failures if a new one occurred, which isn't great either.
> 
> This might be the opportunity to properly fix this. In the old days (pre source coordinators) this wasn't as pressing because global failures were rare and only triggered by the JM (which typically ensured it only happened once), but since this has changed now we should adjust accordingly.

Thanks @zentol  ! Opened https://issues.apache.org/jira/browse/FLINK-34922 to address properly supporting multiple Global failures in the Exception History.

Also changed this PR to ignore Global failures while being in a Restarting/Canceling or Failing phase on the Adaptive scheduler -- let me know what you think

**Pull request closed without merging**